File: install-deps.js

const fs = require("fs").promises;
const path = require("path");
const { exec } = require("child_process");
const { promisify } = require("util");

const execPromise = promisify(exec);

// Mapping of dependency manifest files to their installation commands
const PROJECT_CONFIG = {
    "package.json": "npm install",
    "requirements.txt": "pip install -r requirements.txt",
    "pyproject.toml": "pip install .",
    "Gemfile": "bundle install",
    "go.mod": "go mod download",
};

// Parse command-line arguments
const args = process.argv.slice(2);
const isDryRun = args.includes("--dry-run");
const isVerbose = args.includes("-v") || args.includes("--verbose");
const rootDir = args.find(arg => !arg.startsWith("--")) || ".";

async function installDependencies(dir) {
    if (isVerbose) {
        console.log(`Scanning directory: ${dir}`);
    }

    try {
        const entries = await fs.readdir(dir, { withFileTypes: true });

        for (const entry of entries) {
            const fullPath = path.join(dir, entry.name);

            if (entry.isDirectory()) {
                // Skip common dependency directories
                if (entry.name === "node_modules" || entry.name === ".venv" || entry.name === ".git") {
                    continue;
                }
                await installDependencies(fullPath);
            } else {
                const command = PROJECT_CONFIG[entry.name];
                if (command) {
                    console.log(`[INFO] Detected '${entry.name}' in '${dir}'`);

                    if (isDryRun) {
                        console.log(`[WARNING] DRY RUN: Would run '${command}' in '${dir}'`);
                        continue;
                    }

                    try {
                        console.log(`[INFO] Running '${command}' in '${dir}'`);
                        const { stdout, stderr } = await execPromise(command, { cwd: dir });
                        if (isVerbose && stdout) console.log(`[DEBUG] stdout: ${stdout}`);
                        if (stderr) console.error(`[ERROR] stderr: ${stderr}`);
                        console.log(`[INFO] Successfully installed dependencies for '${dir}'`);
                    } catch (error) {
                        console.error(`[ERROR] Failed to install dependencies in '${dir}'`);
                        console.error(`[ERROR] Command: ${error.cmd}`);
                        console.error(`[ERROR] Exit Code: ${error.code}`);
                        console.error(`[ERROR] Stderr: ${error.stderr}`);
                    }
                }
            }
        }
    } catch (error) {
        console.error(`[ERROR] Failed to read directory '${dir}': ${error.message}`);
    }
}

async function main() {
    console.log(`[INFO] Starting dependency scan in '${path.resolve(rootDir)}'`);
    if (isDryRun) {
        console.warn("[WARNING] Dry run mode is enabled. No packages will be installed.");
    }
    await installDependencies(rootDir);
}

main().catch(error => console.error(`[FATAL] An unexpected error occurred: ${error.message}`));

File: install_deps.py

import os
import subprocess
import argparse
import logging

# Configure logging
logging.basicConfig(
    level=logging.INFO,
    format="[%(levelname)s] %(message)s"
)

# Mapping of dependency manifest files to their installation commands
PROJECT_CONFIG = {
    "package.json": "npm install",
    "requirements.txt": "pip install -r requirements.txt",
    "pyproject.toml": "pip install .",
    "Gemfile": "bundle install",
    "go.mod": "go mod download",
}

def install_dependencies(root_dir, dry_run=False, verbose=False):
    """
    Traverses all subdirectories to find and install dependencies.

    Args:
        root_dir (str): The root directory to start traversal from.
        dry_run (bool): If True, simulate without running install commands.
        verbose (bool): If True, enable detailed logging.
    """
    if verbose:
        logging.getLogger().setLevel(logging.DEBUG)

    logging.info(f"Starting dependency scan in '{root_dir}'...")
    if dry_run:
        logging.warning("Dry run mode is enabled. No packages will be installed.")

    for dirpath, _, filenames in os.walk(root_dir):
        # Skip common dependency directories to speed up the process
        if "node_modules" in dirpath or ".venv" in dirpath or ".git" in dirpath:
            continue

        for config_file, command in PROJECT_CONFIG.items():
            if config_file in filenames:
                project_path = os.path.abspath(dirpath)
                logging.info(f"Detected '{config_file}' in '{project_path}'")

                if dry_run:
                    logging.info(f"DRY RUN: Would run '{command}' in '{project_path}'")
                    continue

                try:
                    logging.info(f"Running '{command}' in '{project_path}'...")
                    subprocess.run(
                        command,
                        shell=True,
                        check=True,
                        cwd=project_path,
                        capture_output=True,
                        text=True,
                    )
                    logging.info(f"Successfully installed dependencies for '{project_path}'")
                except subprocess.CalledProcessError as e:
                    logging.error(f"Failed to install dependencies in '{project_path}'.")
                    logging.error(f"Command: {e.cmd}")
                    logging.error(f"Exit Code: {e.returncode}")
                    logging.error(f"Stdout: {e.stdout}")
                    logging.error(f"Stderr: {e.stderr}")
                except Exception as e:
                    logging.error(f"An unexpected error occurred in '{project_path}': {e}")

def main():
    """Main function to parse arguments and run the installer."""
    parser = argparse.ArgumentParser(
        description="Traverse a monorepo and install dependencies for all detected projects."
    )
    parser.add_argument(
        "--dry-run",
        action="store_true",
        help="Simulate the process without actually installing dependencies."
    )
    parser.add_argument(
        "-v", "--verbose",
        action="store_true",
        help="Enable verbose output for detailed logging."
    )
    parser.add_argument(
        "directory",
        nargs="?",
        default=".",
        help="The root directory of the monorepo to scan. Defaults to the current directory."
    )
    args = parser.parse_args()

    install_dependencies(args.directory, args.dry_run, args.verbose)

if __name__ == "__main__":
    main()
